### PR TITLE
Enable filter by tx type in listaccounthistory rpc method

### PR DESCRIPTION
**Pull request by bvbfan**

### CODE REVIEW

File: src/masternodes/mn_rpc.cpp [hunk 1]
```diff
@@ -2861,6 +2863,7 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
     uint32_t depth = 100;
     bool noRewards = false;
     std::string tokenFilter;
+    uint8_t txType = uint8_t(CustomTxType::None);
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2866]: Can use `static_cast`

Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2866]: Can define as `const auto txType = CustomTxType::None;`

File: src/masternodes/mn_rpc.cpp [hunk 2]
```diff
@@ -2886,6 +2890,13 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
         if (!optionsObj["token"].isNull()) {
             tokenFilter = optionsObj["token"].get_str();
         }
+
+        if (!optionsObj["txtype"].isNull()) {
+            const auto str = optionsObj["txtype"].get_str();
+            if (str.size() == 1) {
+                txType = uint8_t(CustomTxCodeToType(str[0]));
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2897]: Avoid to cast here

File: src/masternodes/mn_rpc.cpp [hunk 3]
```diff
@@ -2902,6 +2913,10 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
             if (height > startKey.blockHeight || (depth <= startKey.blockHeight && (height < startKey.blockHeight - depth)))
                 return true; // continue
 
+            if (txType && txType != category) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2916]: Should explicitly compare with `CustomTxType::None` as `CustomTxType::None != txType`

File: src/masternodes/mn_rpc.cpp [hunk 4]
```diff
@@ -2945,6 +2960,10 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
                 return true; // continue
             }
 
+            if (txType && txType != category) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2963]: Same as above

File: src/masternodes/mn_rpc.cpp (hunk 3 above)
Comment (bvbfan) [reply src/masternodes/mn_rpc.cpp:2916]: If i get it above by `CustomTxType` i should cast here, category is unsigned char

Comment (ShengguangXiao) [reply src/masternodes/mn_rpc.cpp:3074]: Yes, but at least the first element need to do comparison.

Comment (bvbfan) [reply src/masternodes/mn_rpc.cpp:3074]: We know `None` is 0, it's not going to change :) I'll do the change.

**Comment by bvbfan** [comment]
Reword a commit, fix grammar.

File: src/masternodes/mn_rpc.cpp [hunk 5]
```diff
@@ -3032,10 +3040,26 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
...
+            if (str.size() == 1) {
+                txType = CustomTxCodeToType(str[0]);
+            }
+        }
+        if (!optionsObj["limit"].isNull()) {
+            limit = (uint32_t) optionsObj["limit"].get_int64();
+            depth = std::numeric_limits<decltype(depth)>::max();
+        }
+        if (limit == 0) {
+            limit = std::numeric_limits<decltype(limit)>::max();
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:3055]: If the `limint` is max of uint32_t, will the API call timeout?

Comment (bvbfan) [reply src/masternodes/mn_rpc.cpp:3055]: I don't get what timeout you mean?

**Review by ShengguangXiao — APPROVED** [review]

**Review by monstrobishi — APPROVED** [review]

**Comment by monstrobishi** [comment]
Fixed conflicts myself

**Pull request merged**